### PR TITLE
NF sample_mask

### LINKED ISSUE #26: add initial volume suppression to reach steady state
this would be implemented just like scrubbing.

See this issue:
https://github.com/courtois-neuromod/cneuromod_docs/issues/23

### LINKED ISSUE #125: simplify `_optimize_scrub`
the code is hard to follow.

### LINKED ISSUE #132: generate `sample_mask`
both scrubbing and removal of initial time points can lead to remove time samples entirely. 

It would be useful to generate (and document) usage of a `sample_mask_` attribute in the confound object after loading. 
This mechanism is actually necessary to implement correctly the `scrubbing` strategy of Ciric et al. 2017. 

A related issue should be reported in nilearn: the `sample_mask` option is featured at the initialization of a masker, while there is no reason for the time points removal (or even time points #) to be fixed in the life cycle of a masker object. The `sample_mask` argument should rather be passed to the `*transform*` methods, just like `confounds` is.

### PR DESCRIPTION
**Pull request by htwangtw**
**Major change in API**

Closes #132.
Closes #125.
Closes #26. 

### Sample_mask to handle non-steady-state and scrubbing strategy in Ciric 2017
Method `Confounds.load` now returns
* `confounds_`: confound variable with non-steady-state and scrubbed time point removed; major change.
* `sample_mask_`: index of sample to keep; new in this PR.
* `labels_`: unchanged.

### Concerns
In `fMRIprep` confound outputs, non-steady-state and motion outliers are regressors.
Scrubbing methods in Ciric 2017 requires removing the volumes before confound removal through regression. 
The same should be done on non-steady-state volumes as well. 
The current implementation matches Ciric 2017. 
I am wondering if this will confuse the users because now the confounds_ and the image file is not the same length? 
Another concern is nilearn now supports GLM. The first level model class doesn't have  `sample_mask` as in `NiftiMasker`. It would be good if we can accommodate the GLM as well.

### Proposed solutions
The temporary solution without nilearn changing their stuff is as follow:
1. Add an option for `sample_mask`. 
  If it's on, return the `confounds` trimmed and the `sample_mask` to pass to `NiftiMasker`.
  If off, leave the `confounds` untrimmed, keep the outliers as part of the confound regressors, and `sample_mask` kept as it is. 
2. Add docs on how to use the sample_mask feature in `NiftiMasker` and nilearn GLM class. 
3. Propose relevant change to nilearn

WDYT? Or there's a better way to implement the change?

### To-do 
- [x] Alternative API to address `sample_mask`
- [ ] Improve documentation
- [ ] Check ICA-AROMA output
- [x] Propose change in `nilearn` NiftiMasker `sample_mask` implementation (see #132)
- [x] Propose adding `sample_mask` to nilearn GLM first level model class.  
- [ ] Deal with the code quality metrics

### CODE REVIEW

**Comment by pbellec** [comment]
Huge thanks for starting this PR. It is a non-obvious feature, but I believe the only big piece left for `load_confounds` to handle everything generated by fmriprep.

Re the proposed change in API, this is a tough one.
Removing time points really can break a lot of things, and should probably not be made too easy to implement.

The issue is that `confounds` itself depends on whether or not the user wants to use `sample_mask` ...
For example, if the user applies `sample_mask` with the `Scrubbing` strategy, there should not be any scrubbing regressor in the model.

An alternative API would be: 
 * `load`: returns only `confounds`, and `confounds` is *not* meant to be used in conjunction with `sample_mask`. If initial time points are suppressed, then they are added to the confounds using one-hot encoders.
 * `load_mask`: returns `confounds` and `sample_mask`, and the two *have* to be used together. For the `Scrubbing` strategy, this means no scrubbing confounds are added, but they are incorporated in `sample_mask` instead.

Please let me know what you think.

**Comment by htwangtw** [comment]
This makes a lot of a sense and is easier for users to understand! 
It's an easy change in the code, and with this implementation I will not need to rush with the test data. 
I am still not entirely confident with a test case generated from the NKI dataset. 
The reason is it doesn't have fieldmap and single-band references and I am not certain how much that effects the non-steady-state detection. I have never processed data with field map or multiband with fMRIprep, and definitly don't know enough to make a call.

**Comment by pbellec** [comment]
I will chat with @ltetrel because he's selected some data on open-neuro to test fmriprep, and he tried to be able to trigger all the parts of the pipeline. Also I am not sure we can easily redistribute derivatives from NKI. I have not checked in a long time, but from what I remember there is a data usage agreement to be signed to access the dataset.

**Comment by htwangtw** [comment]
Using data from open neuro is a good shout.
The DUA of NKI is only for the phenotype and genetic data. Just to be on the safe side, I checked the imaging data again. The license is no-free (CC-BY-NC) but it does allow redistribution. I would prefer a dataset that's fully open just to be on the safe side.

**Comment by htwangtw** [comment]
I will come back to this PR once we have updated test data - does that sound alright?

**Comment by htwangtw** [comment]
After working on nilearn `sample_mask` issue (see [PR #2857](https://github.com/nilearn/nilearn/pull/2857)), I think we will have a better idea with the API related to sample_mask here.

**Comment by htwangtw** [comment]
Fixed a bug not covered with `test_load_mask`.  
The remaining failed test is related to loading non-steady-state outliers. 
To remain consistent in how the class method `Confounds._load_confound` find different confound strategies, I passed `self` to `Confounds._load_non_steady_state` but not using it at all. 
I prefer to ignore this warning from codacy.

File: load_confounds/tests/test_strategies.py [hunk 1]
```diff
@@ -77,26 +77,27 @@ def test_Scrubbing():
...
         "white_matter_derivative1",
-        "motion_outlier_0",
-        "motion_outlier_1",
     ]
 
     for check in list_check:
         assert check in conf.columns_
+    # should have six motion outliers from scrubbing
+    assert sum("motion_outlier_" in col for col in conf.columns_) == 8
+    # assert len(conf.sample_mask_) == 24
```
Comment (pbellec) [inline load_confounds/tests/test_strategies.py:86]: not sure why this is commented out.

File: load_confounds/tests/test_strategies.py [hunk 2]
```diff
@@ -77,26 +77,27 @@ def test_Scrubbing():
...
 
     # also load confounds with very liberal scrubbing thresholds
     # this should not produce an error
     conf = lc.Scrubbing(fd_thresh=1, std_dvars_thresh=5)
     conf.load(file_confounds)
     assert "motion_outlier_0" not in conf.columns_
+    # assert len(conf.sample_mask_) == 30
```
Comment (pbellec) [inline load_confounds/tests/test_strategies.py:93]: not sure why this is commented out.

File: load_confounds/tests/test_parser.py [hunk 3]
```diff
@@ -400,3 +405,23 @@ def test_ica_aroma():
...
+    reg = conf.load(file_confounds)
+
+    conf_masked = lc.Confounds(strategy=["motion", "scrub"], scrub="full", fd_thresh=0.15)
+    reg_masked, mask = conf_masked.load_mask(file_confounds)
+    # the current test data has 8 time points marked as motion outliers,
+    # and one nonsteady state (overlap with the first motion outlier)
+    assert reg.shape[0] - len(mask) == 8
+    assert reg_masked.shape[0] - len(mask) == 8
+    assert reg.shape[1] - reg_masked.shape[1] == 9
+    # the difference between the two method is in wheather motion outliers
```
Comment (pbellec) [inline load_confounds/tests/test_parser.py:422]: typos "method" -> "methods"; "wheather" -> "whether"

**Review by pbellec — CHANGES_REQUESTED** [review]

**Comment by pbellec** [comment]
I finally found the time to go over the PR. 
I have very little to say: I found the code clear and easy to follow. 
A wonderful contribution!

I've marked a couple of typos and some commented lines in the tests which should either be removed or uncommented, but this is ready to be merged as far as I am concerned. 

My only (potentially) substantial issues are on the API design.

First, it may be more intuitive to have the `load` method send back both `confounds` and `mask_samples` no matter what. If there is no motion outlier and no non-steady volume, the mask would simply retain every volume. We would then have an option for scrubbing, something like `scrub_mask` which would switch between adding scrubbed volumes in `confounds` (`scrub_mask=False`) vs adding them to `sample_masks` (`scrub_mask=True`). We could also not have this parameter at all and offer `scrub_mask=True` as our only option - if we ever implement a `Despiking` strategy, this one proceeds with the regressors - but I think we can live without this alternative strategy and avoid crowding further an already over-crowded API. The `non_steady_state` frames would always get encoded in `sample_mask`. The `load_mask` method would disappear. I realize this is a bit of a 180 degree pivot from my initial feedback on the API. But two things have changed. The main one is that nilearn has implemented the `mask_sample` across all maskers. I wanted to preserve a method that would work without `mask_sample` because nilearn at the time did not support it, but nilearn ended up moving faster than this PR, and this concern is now obsolete :). The other change is that there is now a clear perspective to merge `load_confounds` into nilearn. So now is as good a time as any to break our API: users will discover the final API in nilearn for the first time anyway.

Second, the attribute in the confounds object is `sample_mask_` but the `load_mask` method returns `sample_masks` (with an `s`). I realize this mimics what's going on with `confounds` inside the parser. But I would be in favour of using exactly the same name as the corresponding parameter in the nilearn maskers. `confounds` is meant to be plugged in the `confounds` parameter of the masker, so following this logic the sample mask should be `sample_mask` and not `sample_masks`. I think it only needs to be changed in the docs and the tutorials, but internal names could stay as they are right now (there is an additional suffix `_out` anyway, and `load_single` does use `sample_mask` already). 

Let me know what you think @htwangtw

**Comment by htwangtw** [comment]
Thanks to the super detailed feedbacks @pbellec. I think these are all good points that we can address. I don't have much to add really.

> First, it may be more intuitive to have the `load` method send back both `confounds` and `mask_samples` no matter what. If there is no motion outlier and no non-steady volume, the mask would simply retain every volume. We would then have an option for scrubbing, something like `scrub_mask` which would switch between adding scrubbed volumes in `confounds` (`scrub_mask=False`) vs adding them to `sample_masks` (`scrub_mask=True`). We could also not have this parameter at all and offer `scrub_mask=True` as our only option - if we ever implement a `Despiking` strategy, this one proceeds with the regressors - but I think we can live without this alternative strategy and avoid crowding further an already over-crowded API. The `non_steady_state` frames would always get encoded in `sample_mask`. The `load_mask` method would disappear. 

This is a great idea and simplify the API! The one thing I would add is to check how Nilearn addresses sample_mask or scrubbing in the documentation. I would like to make sure the language is consistent and helpful. For new imager or less method focused researcher, the relationship between `sample_mask`, `scrub_mask`, and the nilearn masker parameter `mask_sample` might not be immediately obvious. It's fine to ignore this point for now I feel. We probably won't know much about how to come up with more consistent descriptors until we have some user feedbacks. Or it might be me over thinking this :P

> I realize this is a bit of a 180 degree pivot from my initial feedback on the API. But two things have changed. The main one is that nilearn has implemented the `mask_sample` across all maskers. I wanted to preserve a method that would work without `mask_sample` because nilearn at the time did not support it, but nilearn ended up moving faster than this PR, and this concern is now obsolete :). The other change is that there is now a clear perspective to merge `load_confounds` into nilearn. So now is as good a time as any to break our API: users will discover the final API in nilearn for the first time anyway.

Sometimes we need to get to an intermediate stage before coming to a better solution. I agree this is a good time to make the change. Can either merge this and then implement the changes here, or just implement this in the same PR

> Second, the attribute in the confounds object is `sample_mask_` but the `load_mask` method returns `sample_masks` (with an `s`). I realize this mimics what's going on with `confounds` inside the parser. But I would be in favour of using exactly the same name as the corresponding parameter in the nilearn maskers. `confounds` is meant to be plugged in the `confounds` parameter of the masker, so following this logic the sample mask should be `sample_mask` and not `sample_masks`. I think it only needs to be changed in the docs and the tutorials, but internal names could stay as they are right now (there is an additional suffix `_out` anyway, and `load_single` does use `sample_mask` already).

Nice catch on this one! Consistent parameter name with nilearn is a good shout.

**Comment by pbellec** [comment]
Happy to hear we are on the same page.

Only follow-up comment: I agree it will be confusing for users to get both `sample_mask` and `scrub_mask`. After sleeping on it, we should probably only offer to scrub through `sample_mask`, as it was the original approach. This will simplify the API and limit possible confusion. The benefits of the regressor-based scrubbing is not obvious to me, so it adds complexity with no tangible benefits.

File: load_confounds/tests/test_strategies.py (hunk 1 above)
Comment (htwangtw) [reply load_confounds/tests/test_strategies.py:86]: In the test, I tested method `load` rather than `load_mask`, so `sample_mask_` was not returned. As we are going to return `sample_mask` as the default for the scrubbing strategy, this test will be rewritten.

File: load_confounds/tests/test_strategies.py (hunk 2 above)
Comment (htwangtw) [reply load_confounds/tests/test_strategies.py:93]: Same as above

**Comment by htwangtw** [comment]
I caught an implementation error of the non-steady-state volume handling while making changes. 

`load_confounds/confounds.py#L270` is not covered by the current test as this is related to the regressor based scrubbing. We are not exposing this feature to users in the current implementation. I would prefer to keep it but let the test coverage slide.. WDYT @pbellec

File: load_confounds/confounds.py [hunk 4]
```diff
@@ -217,24 +216,59 @@ def _confounds_to_df(image_file, flag_acompcor, flag_full_aroma):
...
+    confounds = confounds.values
+    return sample_mask, confounds, labels
+
+
+def _demean_confounds(confounds, outliers, scrub_mask):
+    """Demean the confounds."""
+    confound_cols = confounds.columns
+    confounds = scale(confounds, axis=0, with_std=False)
+    confounds = pd.DataFrame(confounds, columns=confound_cols)
+    # using despike strategy (currently not supported)
```
Comment (pbellec) [inline load_confounds/confounds.py:268]: should we remove this bit of code as it's not supported?

File: load_confounds/confounds.py [hunk 5]
```diff
@@ -217,24 +216,59 @@ def _confounds_to_df(image_file, flag_acompcor, flag_full_aroma):
...
+            confounds = _demean_confounds(confounds, outliers, scrub_mask)
+    labels = list(confounds.columns)
+    confounds = confounds.values
+    return sample_mask, confounds, labels
+
+
+def _demean_confounds(confounds, outliers, scrub_mask):
+    """Demean the confounds."""
+    confound_cols = confounds.columns
+    confounds = scale(confounds, axis=0, with_std=False)
```
Comment (pbellec) [inline load_confounds/confounds.py:266]: we may want to compute the mean only on non-outlier time samples (in a `fit` step, and then apply the demeaning to the entire time series (with a `transform` step).

File: load_confounds/tests/test_parser.py [hunk 6]
```diff
@@ -400,3 +416,28 @@ def test_ica_aroma():
...
+    reg, mask = conf.load(file_confounds)
+
+    # the current test data has 6 time points marked as motion outliers,
+    # and one nonsteady state (overlap with the first motion outlier)
+    # 2 time points removed due to the "full" srubbing strategy
+    assert reg.shape[0] - len(mask) == 8
+    # nilearn requires unmasked confound regressors
+    assert reg.shape[0] == 30
+
+    # non steady state will always be reomoved
```
Comment (pbellec) [inline load_confounds/tests/test_parser.py:435]: typo "reomoved" -> "removed"

**Review by pbellec — APPROVED** [review]

**Comment by pbellec** [comment]
fantastic work @htwangtw I have made super-minor comments, and ready to merge :tada: :tada: :tada:

File: load_confounds/confounds.py (hunk 4 above)
Comment (htwangtw) [reply load_confounds/confounds.py:268]: Yeah let's just remove them

File: load_confounds/confounds.py (hunk 5 above)
Comment (htwangtw) [reply load_confounds/confounds.py]: In the current `NifitiMasker` API, both confounds and the sample mask are passed to the masker at the `transform` step. I think we should just make sure the output from here is demeaned with non-outlier samples only. Might need to discuss with Nilearn side when the merge happens.
